### PR TITLE
[RFC] Use explicit RAML conversion instead of implicits.

### PR DESCRIPTION
**Pull request by jeschkies**
Summary:
This leverages method overloading instead of implicits. This way it is
much easier to find the implementation of a conversion and their usages.
It should also speed up compilation since searching for implicits takes
the longest time.

### CODE REVIEW

**Comment by jeschkies** [comment]
It is not complete yet.

File: src/main/scala/mesosphere/marathon/raml/VolumeConversion.scala [hunk 1]
```diff
@@ -45,7 +45,7 @@ trait VolumeConversion extends ConstraintConversion with DefaultConversions {
         case DiskType.Path => PersistentVolumeType.Path
         case DiskType.Root => PersistentVolumeType.Root
       })
-      PersistentVolume(pvType, pv.size, pv.maxSize, pv.constraints.toRaml[Set[Seq[String]]])
+      PersistentVolume(pvType, pv.size, pv.maxSize, pv.constraints.map(constraintProtosToStringSeq))
```
Comment (timcharper) [inline src/main/scala/mesosphere/marathon/raml/VolumeConversion.scala]: This change should definitely stay

**Comment by timcharper** [comment]
I'm conflicted about how I feel about this. If we were to implement RAML conversion in Rust, it would fromRaml and toRAML would be implemented in a RAMLConversion trait, and then each implementor would define those methods. The benefit is that we get uniformity in definition, and ad-hoc polymorphism. If a method accepts any T for which a RAMLConversion is implemented, then we can simply provide it said T.

So, effectively, I would stand in support for RAMLConversion in the account that it is implemented and treated as a type-class, and the majority of definitions are defined in the companion object for the respective type (this makes for a simple lookup rule for finding implicit definitions).

If a toRaml / fromRAML type needs additional parameters, because normalization is involved, then I'd be inclined to extract that normalization out of the RAML conversion layer (it probably does not belong there). This would make our use of implicits more straightforward, if it is the case that they are currently used this way.

I don't think the implicit class methods are necessary. `.toRaml` etc.

WRT to compile speed, have we run a clean compile benchmark to see if this actually improves things?

File: src/main/scala/mesosphere/marathon/api/akkahttp/EntityMarshallers.scala [hunk 2]
```diff
@@ -17,12 +17,13 @@ import mesosphere.marathon.core.appinfo.AppInfo
...
 import mesosphere.marathon.core.launchqueue.LaunchQueue.QueuedInstanceInfoWithStatistics
 import mesosphere.marathon.core.plugin.PluginDefinitions
+import mesosphere.marathon.raml.AppConversion
 import mesosphere.marathon.state.AppDefinition
 import play.api.libs.json._
 
 import scala.collection.breakOut
 
-object EntityMarshallers {
+object EntityMarshallers extends AppConversion {
```
Comment (meln1k) [inline src/main/scala/mesosphere/marathon/api/akkahttp/EntityMarshallers.scala]: Why extend when we can just import a method?

File: src/main/scala/mesosphere/marathon/api/v2/AppTasksResource.scala [hunk 3]
```diff
@@ -52,13 +50,13 @@ class AppTasksResource @Inject() (
           val groupPath = gid.toRootPath
           val maybeGroup = groupManager.group(groupPath)
           withAuthorization(ViewGroup, maybeGroup, unknownGroup(groupPath)) { group =>
-            ok(jsonObjString("tasks" -> runningTasks(group.transitiveAppIds, instancesBySpec).toRaml))
+            ok(jsonObjString("tasks" -> runningTasks(group.transitiveAppIds, instancesBySpec).map(asRaml)))
```
Comment (meln1k) [inline src/main/scala/mesosphere/marathon/api/v2/AppTasksResource.scala]: Is it the same as before?

File: src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala [hunk 4]
```diff
@@ -445,7 +445,7 @@ trait AppAndGroupFormats {
         info.maybeCounts.map(TaskCountsWrites.writes(_).as[JsObject]),
         info.maybeDeployments.map(deployments => Json.obj("deployments" -> deployments)),
         info.maybeReadinessCheckResults.map(readiness => Json.obj("readinessCheckResults" -> readiness)),
-        info.maybeTasks.map(tasks => Json.obj("tasks" -> Raml.toRaml(tasks))),
+        info.maybeTasks.map(tasks => Json.obj("tasks" -> tasks.map(asRaml))),
```
Comment (meln1k) [inline src/main/scala/mesosphere/marathon/api/v2/json/Formats.scala]: I'm not sure if it's equivalent

File: src/main/scala/mesosphere/marathon/raml/AppConversion.scala [hunk 5]
```diff
@@ -40,7 +37,7 @@ trait AppConversion extends DefaultConversions with ConstraintConversion with En
       backoffFactor = app.backoffStrategy.factor,
       backoffSeconds = app.backoffStrategy.backoff.toSeconds.toInt,
       cmd = app.cmd,
-      constraints = app.constraints.toRaml[Set[Seq[String]]],
+      constraints = app.constraints.map(constraintProtosToStringSeq),
```
Comment (meln1k) [inline src/main/scala/mesosphere/marathon/raml/AppConversion.scala]: Is it the same as it was before?

**Comment by jeschkies** [comment]
> If a toRaml / fromRAML type needs additional parameters, because normalization is involved, then I'd be inclined to extract that normalization out of the RAML conversion layer (it probably does not belong there). This would make our use of implicits more straightforward, if it is the case that they are currently used this way.

+1 on that. Let's revisit this later.

**Pull request closed without merging**